File: bindings/python/bench.py
#!/usr/bin/env python3
"""
Cross-Runtime Comparative Benchmark Suite for Expanse Python Bindings (PyO3).
Compares ExpanseMap and ExpanseSet against Python native dict and set.
"""

from __future__ import annotations

import argparse
import json
import random
import sys
import time
import tracemalloc
from pathlib import Path

# If script directory was auto-prepended to sys.path[0], remove it so Python
# imports the compiled package from site-packages (avoiding source-tree .so lookup failure).
script_dir = str(Path(__file__).resolve().parent)
if sys.path and sys.path[0] == script_dir:
    sys.path.pop(0)

try:
    from expanse_trie import ExpanseMap, ExpanseSet
except ImportError:
    sys.path.insert(0, script_dir)
    try:
        from expanse_trie import ExpanseMap, ExpanseSet
    except ImportError:
        print("Error: expanse_trie module not found. Build with maturin develop or pip install .", file=sys.stderr)
        sys.exit(1)


def parse_args():
    p = argparse.ArgumentParser(description="Expanse Python Bindings Benchmark Suite")
    p.add_argument("--pop", type=int, default=100_000, help="Population size (default: 100,000)")
    p.add_argument("--quick", action="store_true", help="Quick mode (N = 20,000)")
    p.add_argument("--json", action="store_true", help="Emit JSON output for machine parsing")
    return p.parse_args()


def generate_keys(pop: int, dist: str = "random") -> list[int]:
    rng = random.Random(0x0DDB_1A5E_5EED_0001)
    if dist == "sequential":
        return list(range(pop))
    elif dist == "clustered":
        out = []
        base = 0
        for i in range(pop):
            if i % 256 == 0:
                base = rng.getrandbits(64) & ~0xFF
            out.append(base + (i % 256))
        return out
    else:
        return [rng.getrandbits(64) for _ in range(pop)]


def measure(fn, rounds: int = 3) -> float:
    best = float("inf")
    for _ in range(rounds):
        t0 = time.perf_counter()
        fn()
        t1 = time.perf_counter()
        dt = t1 - t0
        if dt < best:
            best = dt
    return best


def run_suite(pop: int, dist: str = "random") -> dict:
    keys = generate_keys(pop, dist)
    probe_keys = keys.copy()
    random.Random(0x9E37_79B9).shuffle(probe_keys)

    # 1. ExpanseMap
    exp_map = ExpanseMap()

    def bench_exp_insert():
        exp_map.clear()
        for k in keys:
            exp_map.insert(k, k ^ 0x55)

    exp_insert_s = measure(bench_exp_insert)

    def bench_exp_lookup():
        sink = 0
        for k in probe_keys:
            v = exp_map.get(k)
            if v is not None:
                sink ^= v
        return sink

    exp_lookup_s = measure(bench_exp_lookup)

    def bench_exp_iter():
        count = 0
        for _k, _v in exp_map.items():
            count += 1
        return count

    exp_iter_s = measure(bench_exp_iter)
    exp_bytes_per_key = float(exp_map.mem_used()) / pop

    # 2. Python dict
    py_dict = {}

    def bench_py_insert():
        py_dict.clear()
        for k in keys:
            py_dict[k] = k ^ 0x55

    py_insert_s = measure(bench_py_insert)

    def bench_py_lookup():
        sink = 0
        for k in probe_keys:
            v = py_dict.get(k)
            if v is not None:
                sink ^= v
        return sink

    py_lookup_s = measure(bench_py_lookup)

    def bench_py_iter():
        count = 0
        for _k, _v in py_dict.items():
            count += 1
        return count

    py_iter_s = measure(bench_py_iter)

    tracemalloc.start()
    sample_dict = {k: k ^ 0x55 for k in keys[:10_000]}
    _cur, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    py_bytes_per_key = float(peak) / 10_000 if peak > 0 else 64.0

    # 3. ExpanseSet
    exp_set = ExpanseSet()

    def bench_exp_set_insert():
        exp_set.clear()
        for k in keys:
            exp_set.insert(k)

    exp_set_insert_s = measure(bench_exp_set_insert)

    def bench_exp_set_lookup():
        count = 0
        for k in probe_keys:
            if exp_set.contains(k):
                count += 1
        return count

    exp_set_lookup_s = measure(bench_exp_set_lookup)

    # 4. Python set
    py_set = set()

    def bench_py_set_insert():
        py_set.clear()
        for k in keys:
            py_set.add(k)

    py_set_insert_s = measure(bench_py_set_insert)

    def bench_py_set_lookup():
        count = 0
        for k in probe_keys:
            if k in py_set:
                count += 1
        return count

    py_set_lookup_s = measure(bench_py_set_lookup)

    to_mops = lambda s: (pop / s) / 1e6
    to_ns = lambda s: (s * 1e9) / pop

    return {
        "dist": dist,
        "pop": pop,
        "expanse_map": {
            "insert_mops": to_mops(exp_insert_s),
            "lookup_mops": to_mops(exp_lookup_s),
            "lookup_ns": to_ns(exp_lookup_s),
            "iter_mops": to_mops(exp_iter_s),
            "bytes_per_key": exp_bytes_per_key,
        },
        "python_dict": {
            "insert_mops": to_mops(py_insert_s),
            "lookup_mops": to_mops(py_lookup_s),
            "lookup_ns": to_ns(py_lookup_s),
            "iter_mops": to_mops(py_iter_s),
            "bytes_per_key": py_bytes_per_key,
        },
        "expanse_set": {
            "insert_mops": to_mops(exp_set_insert_s),
            "lookup_mops": to_mops(exp_set_lookup_s),
            "lookup_ns": to_ns(exp_set_lookup_s),
        },
        "python_set": {
            "insert_mops": to_mops(py_set_insert_s),
            "lookup_mops": to_mops(py_set_lookup_s),
            "lookup_ns": to_ns(py_set_lookup_s),
        },
    }


def render_table(results: list[dict]):
    print("\n================================================================================")
    print("  Expanse Python Bindings Comparative Performance Report")
    print("================================================================================")

    for r in results:
        print(f"\n[ Distribution: {r['dist']} | Population: {r['pop']:,} ]")
        print(f"{'Target':<20} | {'Lookup (ns)':>11} | {'Lookup (Mops)':>13} | {'Insert (Mops)':>13} | {'Iter (Mops)':>11} | {'B/key':>8}")
        print(f"{'-'*20}-+-{'-'*11}-+-{'-'*13}-+-{'-'*13}-+-{'-'*11}-+-{'-'*8}")

        em = r["expanse_map"]
        print(f"{'ExpanseMap':<20} | {em['lookup_ns']:>11.2f} | {em['lookup_mops']:>13.2f} | {em['insert_mops']:>13.2f} | {em['iter_mops']:>11.2f} | {em['bytes_per_key']:>8.2f}")

        pd = r["python_dict"]
        print(f"{'Python native dict':<20} | {pd['lookup_ns']:>11.2f} | {pd['lookup_mops']:>13.2f} | {pd['insert_mops']:>13.2f} | {pd['iter_mops']:>11.2f} | {pd['bytes_per_key']:>8.2f}")

        es = r["expanse_set"]
        print(f"{'ExpanseSet':<20} | {es['lookup_ns']:>11.2f} | {es['lookup_mops']:>13.2f} | {es['insert_mops']:>13.2f} | {'—':>11} | {'—':>8}")

        ps = r["python_set"]
        print(f"{'Python native set':<20} | {ps['lookup_ns']:>11.2f} | {ps['lookup_mops']:>13.2f} | {ps['insert_mops']:>13.2f} | {'—':>11} | {'—':>8}")

    print("\n================================================================================\n")


def main():
    args = parse_args()
    pop = 20_000 if args.quick else args.pop
    dists = ["random", "sequential", "clustered"]
    results = [run_suite(pop, d) for d in dists]

    if args.json:
        print(json.dumps({"runtime": "python", "results": results}, indent=2))
    else:
        render_table(results)


if __name__ == "__main__":
    main()

File: bindings/ruby/bench.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Cross-Runtime Comparative Benchmark Suite for Expanse Ruby Bindings.
# Compares Expanse::Map and Expanse::Set against native Ruby Hash and Set.

require "optparse"
require "json"
require "set"
require "objspace"
require_relative "lib/expanse"

options = { pop: 50_000, quick: false, json: false }
OptionParser.new do |opts|
  opts.banner = "Usage: bench.rb [options]"
  opts.on("--quick", "Run quick benchmark (N=10,000)") { options[:quick] = true; options[:pop] = 10_000 }
  opts.on("--pop N", Integer, "Population size") { |n| options[:pop] = n }
  opts.on("--json", "Emit JSON output") { options[:json] = true }
end.parse!

class XorShift64
  def initialize(seed = 0x0DDB_1A5E_5EED_0001)
    @state = seed & 0xFFFF_FFFF_FFFF_FFFF
  end

  def next
    x = @state
    x ^= (x << 13) & 0xFFFF_FFFF_FFFF_FFFF
    x ^= (x >> 7) & 0xFFFF_FFFF_FFFF_FFFF
    x ^= (x << 17) & 0xFFFF_FFFF_FFFF_FFFF
    @state = x
    x
  end
end

def generate_keys(pop, dist = "random")
  rng = XorShift64.new
  if dist == "sequential"
    Array.new(pop) { |i| i }
  elsif dist == "clustered"
    base = 0
    Array.new(pop) do |i|
      base = rng.next & ~0xFF if (i % 256).zero?
      base + (i % 256)
    end
  else
    Array.new(pop) { rng.next }
  end
end

def measure(rounds = 3)
  best = Float::INFINITY
  rounds.times do
    t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    yield
    t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
    dt = t1 - t0
    best = dt if dt < best
  end
  best
end

def run_suite(pop, dist = "random")
  keys = generate_keys(pop, dist)
  probe_keys = keys.shuffle(random: Random.new(0x9E37_79B9))

  # 1. Expanse::Map
  GC.start
  exp_map = Expanse::Map.new
  exp_insert_s = measure do
    exp_map.clear
    keys.each { |k| exp_map[k] = k ^ 0x55 }
  end

  exp_lookup_s = measure do
    sink = 0
    probe_keys.each do |k|
      v = exp_map[k]
      sink ^= v if v
    end
    sink
  end

  exp_iter_s = measure do
    count = 0
    exp_map.each { |_k, _v| count += 1 }
    count
  end

  # Real native arena accounting is REQUIRED: pre-#373 this silently fell back
  # to hardcoded constants (22.5/8.6), so the nightly Ruby memory gate could
  # structurally never fire.
  unless exp_map.respond_to?(:mem_used)
    abort "Error: Expanse::Map#mem_used is not available — cannot measure bytes/key. " \
          "Rebuild the native library (cargo build --release -p expanse-capi) and update lib/expanse.rb."
  end
  exp_bytes_per_key = exp_map.mem_used.to_f / pop

  # 2. Ruby Hash
  GC.start
  rb_hash = {}
  rb_insert_s = measure do
    rb_hash.clear
    keys.each { |k| rb_hash[k] = k ^ 0x55 }
  end

  rb_lookup_s = measure do
    sink = 0
    probe_keys.each do |k|
      v = rb_hash[k]
      sink ^= v if v
    end
    sink
  end

  rb_iter_s = measure do
    count = 0
    rb_hash.each { |_k, _v| count += 1 }
    count
  end

  # ObjectSpace.memsize_of is a SHALLOW measurement: it covers the Hash's own
  # entry table but not off-slab key/value objects. Random 64-bit keys exceed
  # Ruby's 62-bit Fixnum range, so many keys are separately-allocated Bignums
  # that this number does not include — hence the estimated flag (it is a
  # lower bound, not the fabricated 64.0 constant emitted pre-#373).
  rb_hash_bytes_per_key = ObjectSpace.memsize_of(rb_hash).to_f / pop

  # 3. Expanse::Set
  exp_set = Expanse::Set.new
  exp_set_insert_s = measure do
    exp_set.clear
    keys.each { |k| exp_set.add(k) }
  end

  exp_set_lookup_s = measure do
    count = 0
    probe_keys.each { |k| count += 1 if exp_set.include?(k) }
    count
  end

  # 4. Ruby Set
  rb_set = Set.new
  rb_set_insert_s = measure do
    rb_set.clear
    keys.each { |k| rb_set.add(k) }
  end

  rb_set_lookup_s = measure do
    count = 0
    probe_keys.each { |k| count += 1 if rb_set.include?(k) }
    count
  end

  to_mops = ->(s) { (pop / s) / 1e6 }
  to_ns = ->(s) { (s * 1e9) / pop }

  {
    dist: dist,
    pop: pop,
    expanse_map: {
      insert_mops: to_mops.call(exp_insert_s),
      lookup_mops: to_mops.call(exp_lookup_s),
      lookup_ns: to_ns.call(exp_lookup_s),
      iter_mops: to_mops.call(exp_iter_s),
      bytes_per_key: exp_bytes_per_key
    },
    ruby_hash: {
      insert_mops: to_mops.call(rb_insert_s),
      lookup_mops: to_mops.call(rb_lookup_s),
      lookup_ns: to_ns.call(rb_lookup_s),
      iter_mops: to_mops.call(rb_iter_s),
      bytes_per_key: rb_hash_bytes_per_key,
      # Shallow ObjectSpace.memsize_of / pop: excludes off-slab Bignum keys
      # (random 64-bit keys > 2**62), so this is a lower-bound estimate.
      bytes_per_key_estimated: true
    },
    expanse_set: {
      insert_mops: to_mops.call(exp_set_insert_s),
      lookup_mops: to_mops.call(exp_set_lookup_s),
      lookup_ns: to_ns.call(exp_set_lookup_s)
    },
    ruby_set: {
      insert_mops: to_mops.call(rb_set_insert_s),
      lookup_mops: to_mops.call(rb_set_lookup_s),
      lookup_ns: to_ns.call(rb_set_lookup_s)
    }
  }
end

def render_table(results)
  puts "\n================================================================================"
  puts "  Expanse Ruby Bindings Comparative Performance Report"
  puts "================================================================================"

  results.each do |r|
    puts "\n[ Distribution: #{r[:dist]} | Population: #{r[:pop]} ]"
    printf "%-20s | %11s | %13s | %13s | %11s | %8s\n", "Target", "Lookup (ns)", "Lookup (Mops)", "Insert (Mops)", "Iter (Mops)", "B/key"
    puts "#{'-' * 20}-+-#{'-' * 11}-+-#{'-' * 13}-+-#{'-' * 13}-+-#{'-' * 11}-+-#{'-' * 8}"

    em = r[:expanse_map]
    printf "%-20s | %11.2f | %13.2f | %13.2f | %11.2f | %8.2f\n", "Expanse::Map", em[:lookup_ns], em[:lookup_mops], em[:insert_mops], em[:iter_mops], em[:bytes_per_key]

    rh = r[:ruby_hash]
    printf "%-20s | %11.2f | %13.2f | %13.2f | %11.2f | %8.2f\n", "Ruby Hash", rh[:lookup_ns], rh[:lookup_mops], rh[:insert_mops], rh[:iter_mops], rh[:bytes_per_key]

    es = r[:expanse_set]
    printf "%-20s | %11.2f | %13.2f | %13.2f | %11s | %8s\n", "Expanse::Set", es[:lookup_ns], es[:lookup_mops], es[:insert_mops], "—", "—"

    rs = r[:ruby_set]
    printf "%-20s | %11.2f | %13.2f | %13.2f | %11s | %8s\n", "Ruby Set", rs[:lookup_ns], rs[:lookup_mops], rs[:insert_mops], "—", "—"
  end
  puts "\n================================================================================\n\n"
end

pop = options[:pop]
dists = %w[random sequential clustered]
results = dists.map { |d| run_suite(pop, d) }

if options[:json]
  puts JSON.pretty_generate({ runtime: "ruby", results: results })
else
  render_table(results)
end
